### PR TITLE
FI-507: Automatic token refresh

### PR DESCRIPTION
**Pull request by Jammjammjamm**
Inferno's tests can take some time, so Inferno needs to be able to automatically refresh access tokens before they expire.

**Submitter:**
- [x] This pull request describes why these changes were made
- [x] Internal ticket for this PR: https://oncprojectracking.healthit.gov/support/browse/FI-507
- [x] Internal ticket links to this PR
- [x] Internal ticket is properly labeled (Community/Program)
- [x] Internal ticket has a justification for its Community/Program label
- [x] Code diff has been reviewed for extraneous/missing code
- [x] Tests are included and test edge cases
- [x] Tests/code quality metrics have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: lib/modules/smart/shared_launch_tests.rb [hunk 1]
```diff
@@ -40,7 +40,16 @@ def validate_token_response_contents(token_response, require_expires_in:)
 
         assert token_response_body['access_token'].present?, 'Token response did not contain access_token as required'
 
-        @instance.update(token: token_response_body['access_token'], token_retrieved_at: DateTime.now)
+        expires_in = token_response_body['expires_in']
+        if expires_in.present? # rubocop:disable Style/IfUnlessModifier
+          assert expires_in.is_a?(Numeric), "`expires_in` field is not a number: #{expires_in.inspect}"
```
Comment (arscan) [inline lib/modules/smart/shared_launch_tests.rb]: The requirement behind here is weaker than I'd like:

https://tools.ietf.org/html/rfc6749#section-5.1

"Numerical values are included as JSON numbers."

But the definition of expires_in is never explicitely called numeric.

"RECOMMENDED.  The lifetime in seconds of the access token.  For
         example, the value "3600" denotes that the access token will
         expire in one hour from the time the response was generated.
         If omitted, the authorization server SHOULD provide the
         expiration time via other means or document the default value."

Comment (arscan) [reply lib/modules/smart/shared_launch_tests.rb]: More importantly, this assert needs to move to the end of the test somehow, because now we aren't saving patient_id and the system is blocked from testing any further until this is fixed (because no patient_id)

Comment (arscan) [reply lib/modules/smart/shared_launch_tests.rb]: I think a reasonable fix is to turn this into a warning (because the requirement is slightly vague), which allows the later code to run and the issue is noted.

Comment (Jammjammjamm) [reply lib/modules/smart/shared_launch_tests.rb]: Yeah, this assertion was really here because `token_expires_in` is an integer field and updating it with a non-integer value raises an error. I'll add some logic to prevent that and change this to a warning.
